lib/container.ts

```diff
@@ -207,55 +207,127 @@ export class Container extends (EventEmitter as {
-			// is a directory
-			const strippedPath = matchingDep.sourceIsDirectory
-				? f
-				: f.split(Path.sep).pop()!;
-
-			const toPath = matchingDep.destinationIsDirectory
-				? Path.join(matchingDep.containerPath, strippedPath)
-				: matchingDep.containerPath;
-
-			return {
-				fromPath: f,
-				toPath,
-			};
+				/* istanbul ignore next */
+				if (matchingDep == null) {
+					throw new InternalInconsistencyError(
+						'Could not find matching file for action group',
+					);
+				}
+
+				// TODO: I'm not sure the logic here is actually correct,
+				// specifically the way we build the destination when the containerPath
+				// is a directory
+
+				// The dockerfile class tries to detect if the destination is a directory,
+				// but we can do one better with the actual container, and look directly
+				// at the destination
+				if (!matchingDep.destinationIsDirectory) {
+					matchingDep.destinationIsDirectory = await this.containerPathIsDirectory(
+						matchingDep.containerPath,
+					);
+				}
+
+				// We can also check if the host path is a directory (only if it's not obvious
+				// from the dockerfile)
+				if (!matchingDep.sourceIsDirectory) {
+					matchingDep.sourceIsDirectory = await Container.hostPathIsDirectory(
+						Path.join(this.hostContextPath, f),
+					);
+				}
+
+				const strippedPath = matchingDep.sourceIsDirectory
+					? f
+					: f.split(Path.sep).pop()!;
+
+				const toPath = matchingDep.destinationIsDirectory
+					? Path.join(matchingDep.containerPath, strippedPath)
+					: matchingDep.containerPath;
+
+				return {
+					fromPath: f,
+					toPath,
+				};
+			}),
+		);
+	}
+
+	private containerPathIsDirectory = _.memoize(async (path: string) => {
+		const output = await this.executeCommandInternal([
+			'/bin/sh',
+			'-c',
+			`if [ -d "${path}" ]; then echo 'true'; else echo 'false'; fi`,
+		]);
+		return output.stdout.trim() === 'true';
+	});
+
+	private static hostPathIsDirectory = _.memoize(async (path: string) => {
+		const stat = await fs.lstat(path);
+		return stat.isDirectory();
+	});
+
+	private async executeCommandInternal(
+		command: string[],
+	): Promise<{ stdout: string; stderr: string }> {
+		let stdout = '';
+		let stderr = '';
+
+		const exec = await this.docker.getContainer(this.containerId).exec({
+			Cmd: command,
+			AttachStderr: true,
+			AttachStdout: true,
 		});
+
+		return await new Promise<{ stdout: string; stderr: string }>(
+			(resolve, reject) => {
+				exec.start(async (err: Error, stream: Stream.Readable) => {
+					if (err) {
+						reject(err);
+					}
+
+					const stdoutStream = new Stream.PassThrough();
+					const stderrStream = new Stream.PassThrough();
+					stream.on('end', async () => {
+						stdoutStream.end();
+						stderrStream.end();
+						resolve({ stdout, stderr });
+					});
+					this.docker.modem.demuxStream(stream, stdoutStream, stderrStream);
+					stderrStream.on('data', d => {
+						stderr += d.toString();
```

Comment:
It would be better to do:
```ts
const stdout = [];
```
```ts
stdout.push(d)
```
```ts
Buffer.concat(stdout)
```
so as to avoid the constant conversion to strings/concatenation